{% extends "ptp/base3.html" %}
{% block forms %}
<p>
<b>1. What is PTP and bPTP?</b>
</p>
<p>
PTP can give
species delimitation hypothesis based on a gene trees inferred from 
molecular sequences. PTP stands for the Poisson Tree 
Processes model. In PTP, we model speciations or branching events 
in terms of number of mutations. So it only requires a phylogenetic 
input tree, for example the output of RAxML - the 
branch lengths should represent number of mutations. Our numerous tests
show PTP outperforms GMYC. More importantly, PTP is very easy to use, 
since it can use the phylogenetic tree directly without the 
difficult and error prone procedures of time calibration required by 
GMYC. bPTP is a Bayesian implementation of the PTP model.
</p>

<p>
<b>2. How can I get the phylogenetic tree?</b>
</p>
<p>
Google "RAxML" or "Phyml" or "Mrbayes".
Want to use a web server? Try:
<a href = http://embnet.vital-it.ch/raxml-bb/>http://embnet.vital-it.ch/raxml-bb/</a>
or 
<a href = http://www.atgc-montpellier.fr/phyml>http://www.atgc-montpellier.fr/phyml/</a>
</p>
<p>
The server will only accept Newick format or NEXUS format with no annotations on the tree, the tree should be strictly bifurcating.<br />
Here is an <a href = "/static/img/example.tre">example of Newick format</a> and an <a href = "/static/img/example.nex">example of NEXUS format</a>.
</p>

<p>
<span style="color:red">
<b>Special note on taxa name:</b>
A valid taxa name should only contains letters, numbers, _ and -; and should NOT be pure numbers. 
Any other characters, in particular space, $, #, %, ( and ),  will cause errors! 
</span>
</p>

<p>
<b>3. How many generations of MCMC should I run?</b>
</p>
<p>
As long as possible, but the server has a limit of 500,000 generations.
For small trees (<50 taxa), 100,000 generations is
usually enough. For large trees, you should run the MCMC analysis much longer,
and always check for convergence (I will talk about this in the following).
If your tree is larger, say a few hundred taxa, you can start with 100,000 generations
and the analysis will be fast (< 30 min), if it is not enough, change the seed and run 
longer. In case you have very large trees - more than a thousand taxa, you should
download the stand alone version of bPTP and try 1 million generations at least. 
Alternatively, you can use the maximal likelihood solution, see below. 
</p>

<p>
<b>4. Why should I always, always check for convergence? </b>
</p>
<p>
The answer is simple, if the MCMC chains did not converge, the results are wrong. 
All those Bayesian support values will be meaningless if the MCMC chains did not 
reach the equilibrium distribution.
</p>

<p>
<b>
5. How do I check for convergence?
</b>
</p>
<p>
We only care about the species delimitation, so visual checking the likelihood plot
of each delimitation is sufficient. Due to the nature of this model and my implementation, 
mixing is not an issue. Upon convergence, the chain should stay at high 
likelihood locations most of the time and some times explores low likelihood locations.
I will show you two examples so you can have an intuitive feelings about it.
<br \><br \>
A typical example of a converged MCMC chain:<br \>
<img src="/static/img/converge.png" />

<br \>
A typical example of a NOT converged MCMC chain:<br \>
<img src="/static/img/no_converge.png" />
</p>

<p>
<b>
6. What if my MCMC chain does not converge no matter how I try?
</b>
</p>
<p>
This could happen a. your tree is really large and the server has a limit on number of MCMC runs. 
b. PTP is not a good model for your data.
c. The likelihood surface is really rough and you had bad lucks.
If getting the Bayesian results is frustrating, remember you can always fall back to the maximal likelihood 
solution. A flat prior is given to each possible delimitations, so if a single tree is used, 
there is always a maximal likelihood solution. However, remember Bayesian supports 
for delimited species is meaningless without a converged MCMC chain.
</p>

<p>
<b>
7. What if I do not like Bayesian approach?
</b>
</p>
<p>
See question 6. Note there is also a bootstrap version of PTP in my <a href = "https://github.com/zhangjiajie/SpeciesCounting">GitHub</a>. 
</p>

<p>
<b>
8. How can I get the maximal likelihood solution?
</b>
</p>
<p>
See question 6.
</p>


<p>
<b>
8. How long does it take for the analysis to finish?
</b>
</p>
<p>
This depends on your tree size and shape, plus the MCMC generation you specified.
30 taxa and 100,000 generations needs only 1 min, 
300 taxa and 500,000 generations might take no more than a few hours. 
1000 taxa and 500,000 generations might take days and 5000 taxa might take weeks.
</p>



<p>
<b>
9. What does Bayesian support value mean?
</b>
</p>
<p>
Support values shown on the tree plot are computed as the "number of 
occurrence of all the descendants under this node"/ "number of samples
from MCMC sampling". They are the posterial probabilities of those taxa 
form one species under the PTP model and a flat prior. From tests on 
simulated data, support values are strongly correlated with the accuracy 
of the delimitation, r = 0.91.  
</p>


<p>
<b>
10. How do I cite PTP and bPTP?
</b>
</p>
<p>
Please cite: 
        <a href = "http://bioinformatics.oxfordjournals.org/content/29/22/2869">
        <b>A General Species Delimitation Method with Applications to Phylogenetic Placements.</b> 
        <b>Zhang, Jiajie, Kapli, P., Pavlidis, P., and Stamatakis, A.</b> 
        <i>Bioinformatics (Oxford, England)(2013), 29 (22): 2869-2876</i> 
        </a> 
</p>


<p>
<b>
11. Questions, bugs and suggestions?
</b>
</p>
<p>
Please post them on the <a href = "https://groups.google.com/forum/?hl=en#!forum/ptp-species-delimitation">Google group </a>.
If you think there is a bug, please e-mail me your job id, e-mail address, the parameters you used to run the job, and attatch your input tree file.
</p>

{% endblock %}
